Support tensor subclasses in Torchscript

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54817 Support tensor subclasses in Torchscript**

Summary:

Test Plan: python test case

Reviewers: gmagogsfm

Subscribers:

Tasks: #53933

Tags:

Differential Revision: [D27407723](https://our.internmc.facebook.com/intern/diff/D27407723)

tugsbayasgalan committed:
Support tensor subclasses in Torchscript

Summary:

Test Plan: python test case

Reviewers: gmagogsfm

Subscribers:

Tasks: #53933

Tags:

[ghstack-poisoned]

gmagogsfm commented on test/test_jit.py:
```diff
@@ -3513,6 +3513,27 @@ def __init__(self):
+        def check_subclass_warn(input: torch.LongTensor) -> torch.LongTensor:
+            return torch.zeros((input.shape[0], 1), dtype=input.dtype)
+
+        with warnings.catch_warnings(record=True) as warns:
```
TIL!

gmagogsfm commented on torch/csrc/jit/frontend/script_type_parser.cpp:
```diff
@@ -193,9 +193,10 @@ c10::optional<std::string> ScriptTypeParser::parseBaseTypeName(
-        return "Tensor";
+      // Special case for torch.Tensor and its' subclasses
+      if (isTorch(select.value()) &&
+          (name.find("Tensor") != std::string::npos)) {
```
This isn't strict enough I think, something like `torch.xTensory` would pass this check as well. Since we are already hard coding names of these tensor classes in `string_to_type.cpp`, can we use an exhaustive list to check for here as well?

tugsbayasgalan replied:
done

gmagogsfm commented on torch/jit/annotations.py:
```diff
@@ -275,11 +276,27 @@ def get_enum_value_type(e: Type[enum.Enum], loc):
+    if issubclass(ann, (torch.LongTensor, torch.DoubleTensor, torch.FloatTensor,
+                        torch.IntTensor, torch.ShortTensor, torch.HalfTensor,
+                        torch.CharTensor, torch.ByteTensor, torch.BoolTensor)):
+        warnings.warn("TorchScript will ignore subtypes of Tensor "
+                      "and does not enforce dtype in compilation. "
+                      "User annotations might be ignored.")
```
```suggestion
        warnings.warn("TorchScript will treat type annotations of Tensor dtype-specific subtypes as if they are normal Tensors. dtype constraints are not enforced in compilation either.")
```

tugsbayasgalan replied:
done

tugsbayasgalan committed:
Update on "Support tensor subclasses in Torchscript"


Summary:

Test Plan: python test case

Reviewers: gmagogsfm

Subscribers:

Tasks: #53933

Tags:

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/frontend/string_to_type.cpp:
```diff
@@ -6,6 +6,15 @@ using namespace c10;
 const std::unordered_map<std::string, TypePtr>& string_to_type_lut() {
   static std::unordered_map<std::string, TypePtr> map = {
       {"Tensor", TensorType::get()},
+      {"LongTensor", TensorType::get()},
```
I think we should leave a comment here explaining why `dtype` is kind of abandoned here.

tugsbayasgalan replied:
done

gmagogsfm commented on torch/csrc/jit/frontend/script_type_parser.cpp:
```diff
@@ -209,9 +209,19 @@ c10::optional<std::string> ScriptTypeParser::parseBaseTypeName(
+      // Special case for torch.Tensor and its' subclasses
+      const std::unordered_set<std::string> tensor_subtypes = {
+          "Tensor",
+          "LongTensor",
```
"DoubleTensor" is missing from this list. Is it intended?

tugsbayasgalan replied:
done

tugsbayasgalan committed:
Update on "Support tensor subclasses in Torchscript"


Summary:

Test Plan: python test case

Reviewers: gmagogsfm

Subscribers:

Tasks: #53933

Tags:

Differential Revision: [D27407723](https://our.internmc.facebook.com/intern/diff/D27407723)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor subclasses in Torchscript"


Summary:

Test Plan: python test case

Reviewers: gmagogsfm

Subscribers:

Tasks: #53933

Tags:

Differential Revision: [D27407723](https://our.internmc.facebook.com/intern/diff/D27407723)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor subclasses in Torchscript"


Summary:

Test Plan: python test case

Reviewers: gmagogsfm

Subscribers:

Tasks: #53933

Tags:

Differential Revision: [D27407723](https://our.internmc.facebook.com/intern/diff/D27407723)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support tensor subclasses in Torchscript"


Summary:

Test Plan: python test case

Reviewers: gmagogsfm

Subscribers:

Tasks: #53933

Tags:

Differential Revision: [D27407723](https://our.internmc.facebook.com/intern/diff/D27407723)

[ghstack-poisoned]
